From patchwork Wed Dec 12 01:54:02 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: gchen.guomin@gmail.com
X-Patchwork-Id: 10725505
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 9F55215A6
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 12 Dec 2018 01:54:34 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 8D5832A56E
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 12 Dec 2018 01:54:34 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 7E8CA2A76D; Wed, 12 Dec 2018 01:54:34 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 296352A56E
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 12 Dec 2018 01:54:34 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726246AbeLLBy1 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Tue, 11 Dec 2018 20:54:27 -0500
Received: from mail-pg1-f194.google.com ([209.85.215.194]:38830 "EHLO
        mail-pg1-f194.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726201AbeLLBy1 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 11 Dec 2018 20:54:27 -0500
Received: by mail-pg1-f194.google.com with SMTP id g189so7543623pgc.5;
        Tue, 11 Dec 2018 17:54:26 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id;
        bh=wMiU8m4FxK5Mmyx9m3CouY8SU7XPJqd6xuSRWNHMlC0=;
        b=RbbAB/tJGzIvFWL6BzgPv49EKviGxgOa5dyIiHKU0Ff/aqpFBFQEL3VKNvpF+wQUvZ
         09yNTwnO21L0j2dTe6/xS87TsS5USWWxr82FFSyQYN6UwPgD0FIxz+MvYIecL0S6mwFn
         yTo5oHWouy0mjkhOFktBMtzYo7pInKuqdMw1jCvYQ3GunaUqtEYhCm4AiiNTDkVGXmas
         5OlNMPOzR5Yrx05BX6gganBGscMUAdBqgnCUyOEnEeWmj6wPxxDYRwj6iIu3YnsV3gqJ
         GURMWmtGvFSbN3651UDuQ/wrwIRaIuX1IMOFk2dFLvApH4lj2W0bCT+nOGRFfdlXVb3z
         4NbQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=wMiU8m4FxK5Mmyx9m3CouY8SU7XPJqd6xuSRWNHMlC0=;
        b=GAiWVKz/o0TPW2rX20M3RAV2neXBZVTdCyqM+VlRQY9tEKXgtcRR/tiExRcmyfxeXl
         0U5ycYG1NK0YrH+DhVHC8JwQ0iqYSvzxPzcOoOjqMIn7oHGg1VodioaTpuiFwgdn4BzB
         8YOnSltd52WwtmvD7HZTgYKA2KK4ngfDqFaQhkfKZrf4kNh0BiWZrTVpqIXI1bh7iicX
         I+dY7rmwd7hLYKrHcRPehectI9priiTZvxSTcCby/9Si3VqPsE7ShUjrNY36EXwG8276
         iy9Ja3orEttDIaPT497Um8vaBKKve8dqJe6K2idnhfYjaaLk9lHfo9podnscBjPfVKh4
         P+7A==
X-Gm-Message-State: AA+aEWaEPmu4wLbVGFXfJG3rhCCYAqfyx/i84N//+DkC5bA9tHUqim9+
        M2cV/fdC0joFfJ6y2E8RbZM=
X-Google-Smtp-Source: 
 AFSGD/WkkVhDjGPIxDxaxnO+6+qESCBRXgjdab7i8B6Uu1lWvSVNFS4nqSooGwbXU0YRW5rbOGHEqw==
X-Received: by 2002:a62:3948:: with SMTP id
 g69mr18795627pfa.114.1544579666117;
        Tue, 11 Dec 2018 17:54:26 -0800 (PST)
Received: from VM_15_152_centos.localdomain ([193.112.178.48])
        by smtp.gmail.com with ESMTPSA id
 r130sm30478813pfr.48.2018.12.11.17.54.23
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 11 Dec 2018 17:54:25 -0800 (PST)
From: gchen.guomin@gmail.com
To: <guominchen@tencent.com>, "Michael S. Tsirkin" <mst@redhat.com>,
        Jason Wang <jasowang@redhat.com>
Cc: guomin chen <gchen.guomin@gmail.com>, <kvm@vger.kernel.org>,
        <virtualization@lists.linux-foundation.org>,
        <netdev@vger.kernel.org>, <linux-kernel@vger.kernel.org>
Subject: [PATCH]  Fix mm->owner point to a task that does not exists.
Date: Wed, 12 Dec 2018 09:54:02 +0800
Message-Id: <1544579642-29626-1-git-send-email-gchen.guomin@gmail.com>
X-Mailer: git-send-email 1.8.3.1
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: guomin chen <gchen.guomin@gmail.com>

 Under normal circumstances,When do_exit exits, mm->owner will
 be updated on exit_mm(). but when the kernel process calls
 unuse_mm() and then exits,mm->owner cannot be updated. And it
 will point to a task that has been released.

 Below is my issue on vhost_net:
    A, B are two kernel processes(such as vhost_worker),
    C is a user space process(such as qemu), and all
    three use the mm of the user process C.
    Now, because user process C exits abnormally, the owner of this
    mm becomes A. When A calls unuse_mm and exits, this mm->ower
    still points to the A that has been released.
    When B accesses this mm->owner again, A has been released.

 Process A             Process B
 vhost_worker()        vhost_worker()
  ---------             ---------
  use_mm()              use_mm()
   ...
  unuse_mm()
     tsk->mm=NULL
   do_exit()            page fault
    exit_mm()           access mm->owner
   can't update owner   kernel Oops

                        unuse_mm()

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Jason Wang <jasowang@redhat.com>
Cc: <kvm@vger.kernel.org>
Cc: <virtualization@lists.linux-foundation.org>
Cc: <netdev@vger.kernel.org>
Cc: <linux-kernel@vger.kernel.org>
Signed-off-by: guomin chen <gchen.guomin@gmail.com>
---
 drivers/vhost/vhost.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 6b98d8e..7f8586a 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -368,6 +368,10 @@ static int vhost_worker(void *data)
 		}
 	}
 	unuse_mm(dev->mm);
+	/* current->mm needs to be processed by the later exit_mm() */
+	task_lock(current);
+	current->mm = dev->mm;
+	task_unlock(current);
 	set_fs(oldfs);
 	return 0;
 }
